diff --git a/fileset.c b/fileset.c
index 6e570f2..5d678c0 100644
--- a/fileset.c
+++ b/fileset.c
@@ -117,6 +117,7 @@ fileset_resolvepath(filesetentry_t *entry)
 	char path[MAXPATHLEN];
 	char pathtmp[MAXPATHLEN];
 	char *s;
+	size_t len;
 
 	path[0] = '\0';
 	while (fsep->fse_parent) {
@@ -127,8 +128,9 @@ fileset_resolvepath(filesetentry_t *entry)
 		fsep = fsep->fse_parent;
 	}
 
-	s = malloc(strlen(path) + 1);
-	(void) fb_strlcpy(s, path, MAXPATHLEN);
+	len = strlen(path) + 1;
+	s = malloc(len);
+	(void) fb_strlcpy(s, path, len);
 	return (s);
 }

